### PR TITLE
Changes to add support for AMETEK Powervar UPM as usb-hid UPS.

### PR DESCRIPTION
**Pull request by andrewmccartney**
Includes adding a report index to libusb.c in order to handle the composite device in the UPM. 

The original code has the report index hard coded to zero. By using a variable for the report index it can be set to the desired report index based on detecting connected devices that use multiple report indexes.

All other changes are the standard items for introducing a new HID device.

### CODE REVIEW

**Comment by clepple** [comment]
> The BUILD_TYPE=default failed for a missing dependency in "make install". I don't think any of my changes could have caused this, any ideas why the build failed like this?

@jimklimov do you have a way to retry that build on Travis?

**Comment by aquette** [comment]
@clepple theoretically you can access by clicking the link and restart build (which I just did)

**Comment by andrewmccartney** [comment]
Thank you to @jimklimov and @clepple for getting the build issue sorted out. Now that it builds without conflicts is there anything else I need to do in order to get the changes merged into the master?

File: drivers/powervar-hid.c
Comment (jimklimov) [inline drivers/powervar-hid.c:7]: @andrewmccartney : I believe your (C) belongs at least here ;)

**Review by jimklimov — APPROVED** [review]
Can't comment on actual driver content (mostly it is HID mapping... and whether correct flags, string-names and other fields are used), but structurally the change looks good :)

Note this will slightly conflict (at least in Makefile and likely in subdriver list) with changes from PR #807

File: drivers/powervar-hid.c
Comment (aquette) [inline drivers/powervar-hid.c:70]: same comments as #807 on indexed string (mfr, model, serial)

File: drivers/powervar-hid.c
Comment (aquette) [inline drivers/powervar-hid.c:95]: @andrewmccartney are these all the usage paths that you get when running usbhid-ups -DDD -x explore
see also #807 on settings (battery.charge.low for ex) and commands

Comment (andrewmccartney) [reply drivers/powervar-hid.c:95]: @aquette Yes, these are the items we are supporting at this time. I don't think we have any commands or settings at this time. Please advise if I'm mis-understanding your comment about settings and commands.

File: drivers/powervar-hid.c
Comment (andrewmccartney) [reply drivers/powervar-hid.c:70]: @aquette Are you saying that we don't need to have the iProduct, iSerialNumber & iManufacturer entries?

Comment (aquette) [reply drivers/powervar-hid.c:70]: Indeed

File: drivers/powervar-hid.c
Comment (aquette) [reply drivers/powervar-hid.c:95]: At least the shutdown and startup delays usages should be implemented to reach the UPS power off feature.
I'd still like to see the debut output

Comment (andrewmccartney) [reply drivers/powervar-hid.c:95]: Here is the debug output as requested

 228.233213	[D3] usb_busses=0x559d6bd44e60
 228.233229	[D2] Checking device (4234/0002) (002/075)
 228.234544	[D2] - VendorID: 4234
 228.234566	[D2] - ProductID: 0002
 228.234573	[D2] - Manufacturer: AMETEK-POWERVAR Corporation
 228.234580	[D2] - Product: UPM UPS (v01.51, Nov  3 2018)
 228.234586	[D2] - Serial Number: 5014403R-1740135
 228.234619	[D2] - Bus: 002
 228.234634	[D2] - Device release number: 0002
 228.234644	[D2] Set rep index to 1
 228.234658	[D2] Trying to match device
 228.234675	[D2] Device matches
 228.234691	[D2] failed to claim USB device: could not claim interface 1: Device or resource busy
 228.234760	[D2] detached kernel driver from USB device...
 228.234787	[D3] nut_usb_set_altinterface: skipped usb_set_altinterface(udev, 0)
 228.235082	[D3] Report[get]: (2 bytes) => 02 02
 228.235111	[D2] Path: UPS.PowerSummary.iProduct, Type: Feature, ReportID: 0x02, Offset: 0, Size: 8, Value: 2
 228.235331	[D3] Report[get]: (2 bytes) => 03 04
 228.235359	[D2] Path: UPS.PowerSummary.iSerialNumber, Type: Feature, ReportID: 0x03, Offset: 0, Size: 8, Value: 4
 228.235373	[D3] Report[buf]: (2 bytes) => 02 02
 228.235387	[D2] Path: UPS.PowerSummary.iMnaufacturer, Type: Feature, ReportID: 0x02, Offset: 0, Size: 8, Value: 2
 228.235582	[D3] Report[get]: (2 bytes) => 04 06
 228.235609	[D2] Path: UPS.PowerSummary.iDeviceChemistry, Type: Feature, ReportID: 0x04, Offset: 0, Size: 8, Value: 6
 228.236330	[D3] Report[get]: (2 bytes) => 06 01
 228.236357	[D2] Path: UPS.PowerSummary.iOEMInformation, Type: Feature, ReportID: 0x06, Offset: 0, Size: 8, Value: 1
 228.237057	[D3] Report[get]: (2 bytes) => 07 02
 228.237076	[D2] Path: UPS.PowerSummary.CapacityMode, Type: Feature, ReportID: 0x07, Offset: 0, Size: 8, Value: 2
 228.237307	[D3] Report[get]: (2 bytes) => 0c 64
 228.237326	[D2] Path: UPS.PowerSummary.RemainingCapacity, Type: Feature, ReportID: 0x0c, Offset: 0, Size: 8, Value: 100
 228.237558	[D3] Report[get]: (2 bytes) => 09 64
 228.237576	[D2] Path: UPS.PowerSummary.FullChargeCapacity, Type: Feature, ReportID: 0x09, Offset: 0, Size: 8, Value: 100
 228.238248	[D3] Report[get]: (2 bytes) => 08 64
 228.238272	[D2] Path: UPS.PowerSummary.DesignCapacity, Type: Feature, ReportID: 0x08, Offset: 0, Size: 8, Value: 100
 228.238418	[D3] Report[get]: (4 bytes) => 0e 40 7e 05
 228.238434	[D2] Path: UPS.PowerSummary.RuntimeToEmpty, Type: Feature, ReportID: 0x0e, Offset: 0, Size: 24, Value: 360000
 228.238686	[D3] Report[get]: (3 bytes) => 0a 04 09
 228.238712	[D2] Path: UPS.PowerSummary.PresentStatus.ACPresent, Type: Feature, ReportID: 0x0a, Offset: 8, Size: 1, Value: 1
 228.238729	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238744	[D2] Path: UPS.PowerSummary.PresentStatus.Discharging, Type: Feature, ReportID: 0x0a, Offset: 13, Size: 1, Value: 0
 228.238758	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238769	[D2] Path: UPS.PowerSummary.PresentStatus.Charging, Type: Feature, ReportID: 0x0a, Offset: 12, Size: 1, Value: 0
 228.238784	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238797	[D2] Path: UPS.PowerSummary.PresentStatus.ShutdownImminent, Type: Feature, ReportID: 0x0a, Offset: 1, Size: 1, Value: 0
 228.238810	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238822	[D2] Path: UPS.PowerSummary.PresentStatus.BelowRemainingCapacityLimit, Type: Feature, ReportID: 0x0a, Offset: 10, Size: 1, Value: 0
 228.238837	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238848	[D2] Path: UPS.PowerSummary.PresentStatus.Overload, Type: Feature, ReportID: 0x0a, Offset: 3, Size: 1, Value: 0
 228.238860	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238874	[D2] Path: UPS.PowerSummary.PresentStatus.NeedReplacement, Type: Feature, ReportID: 0x0a, Offset: 15, Size: 1, Value: 0
 228.238897	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238910	[D2] Path: UPS.PowerSummary.PresentStatus.InternalFailure, Type: Feature, ReportID: 0x0a, Offset: 0, Size: 1, Value: 0
 228.238924	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238936	[D2] Path: UPS.PowerSummary.PresentStatus.Good, Type: Feature, ReportID: 0x0a, Offset: 2, Size: 1, Value: 1
 228.238950	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238964	[D2] Path: UPS.PowerSummary.PresentStatus.OverTemperature, Type: Feature, ReportID: 0x0a, Offset: 4, Size: 1, Value: 0
 228.238977	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238989	[D2] Path: UPS.PowerSummary.PresentStatus.FullyCharged, Type: Feature, ReportID: 0x0a, Offset: 11, Size: 1, Value: 1
 228.239001	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.239013	[D2] Path: UPS.PowerSummary.PresentStatus.FullyDischarged, Type: Feature, ReportID: 0x0a, Offset: 14, Size: 1, Value: 0
 228.239025	[D1] Not using interrupt pipe...
 228.239037	[D1] Quick update...
 228.239049	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.239060	[D2] Path: UPS.PowerSummary.PresentStatus.ACPresent, Type: Feature, ReportID: 0x0a, Offset: 8, Size: 1, Value: 1
 228.239072	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.239082	[D2] Path: UPS.PowerSummary.PresentStatus.Discharging, Type: Feature, ReportID: 0x0a, Offset: 13, Size: 1, Value: 0
 228.239092	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.239104	[D2] Path: UPS.PowerSummary.PresentStatus.Charging, Type: Feature, ReportID: 0x0a, Offset: 12, Size: 1, Value: 0
 228.239118	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.239130	[D2] Path: UPS.PowerSummary.PresentStatus.BelowRemainingCapacityLimit, Type: Feature, ReportID: 0x0a, Offset: 10, Size: 1, Value: 0
 230.234654	[D1] upsdrv_updateinfo...

File: drivers/powervar-hid.c
Comment (andrewmccartney) [reply drivers/powervar-hid.c:70]: Understood. I have removed them and diid a build and an install locally and all is well. I'll wait to upload until we have completed the other review items.

File: drivers/powervar-hid.c
Comment (aquette) [reply drivers/powervar-hid.c:95]: Thx.
Indeed, none of the standard delay to power off the unit. I may have missed something, but are you manufacturing it? If so, how do you achieve the power off in the end?
Otherwise looks good and can be merged

Comment (andrewmccartney) [reply drivers/powervar-hid.c:95]: With regards to the shutdown and startup delays. At this time we have not implemented them in our UPS for HID-battery, so I am assuming that it does not make sense to include them in our NUT driver. Is this correct?

Comment (andrewmccartney) [reply drivers/powervar-hid.c:95]: Looks like our comments were made at the same time ... we do have various shutdown controls in the UPS (timed, RS232 shutdown pin, load based, and of course low battery voltage) but none of them are included in our HID-battery implementation at this time. So if you shutdown the system that is connected to the UPS it will continue to idle on battery until either AC returns or the battery is exhausted, unless you have configured one of the shutdown mechanisms I listed above outside of the HID-battery interface. For example, configure the UPS to shutdown if on battery and the load falls below a minimum level or a timed shutdown from when it transfers to battery. Hopefully this makes sense.

Comment (aquette) [reply drivers/powervar-hid.c:95]: Indeed, so good to me.
And whenever you add FW features, don't forget to update your nut subdriver

File: drivers/powervar-hid.c
Comment (aquette) [reply drivers/powervar-hid.c:70]: Ok

**Comment by jimklimov** [comment]
There were some good comments in discussion about ensuring an UPS poweroff with some clever scenarios, which currently should be set up outside the USB HID connection. I think the manpage needs to mention that, perhaps referencing the vendor documentation (if there are some stable URLs possible) or spelling out the interesting parts in NUT docs/ codebase.

This also brought to my attention that the PR does not update `docs/man/usbhid-ups.txt` to mention the subdriver support being added.

- [ ] Update the docs about new subdriver

**Comment by jimklimov** [comment]
One more point came up, it is worth mentioning the vendors' efforts at https://github.com/networkupstools/nut/blob/master/docs/acknowledgements.txt

File: drivers/powervar-hid.c
Comment (andrewmccartney) [reply drivers/powervar-hid.c:70]: UPSC output with all changes included.

amccartney@amccartney-OptiPlex-3010:/lib/nut$ upsc UPM@localhost
Init SSL without certificate database
battery.capacity: 100
battery.capacitymode: 2
battery.charge: 100
battery.designcapacity: 100
battery.ioeminformation: AMETEK-POWERVAR Cor
battery.runtime: 360000
battery.type: PbAc
device.mfr: AMETEK-POWERVAR Corporation
device.model: UPM UPS (v01.51, Nov  3 2018)
device.serial: 5014403R-1740135
device.type: ups
driver.flag.pollonly: enabled
driver.name: usbhid-ups
driver.parameter.pollfreq: 30
driver.parameter.pollinterval: 2
driver.parameter.port: auto
driver.parameter.synchronous: no
driver.version: 2.7.4-732-gadb504d0
driver.version.data: Powervar HID 0.19
driver.version.internal: 0.43
ups.mfr: AMETEK-POWERVAR Corporation
ups.model: UPM UPS (v01.51, Nov  3 2018)
ups.powersummary.imanufacturer: 2
ups.powersummary.iproduct: 2
ups.powersummary.iserialnumber: 4
ups.productid: 0002
ups.serial: 5014403R-1740135
ups.status: OL
ups.vendorid: 4234

**Comment by jimklimov** [comment]
Thanks for getting to the docs too! Dictionary wants an update per https://travis-ci.org/github/networkupstools/nut/jobs/735038633#L870 :

````
  ASPELL   Spell checking on usbhid-ups.txt
FAILED : Aspell reported errors here:
----- vvv
342:& AMETEK 23 8: ATEK, AMMETER, AMTRAK, EMETIC, ELTEK, AMMETERS, AMT, AMOK, AMATEUR, UNITEK, AMATI, AMIDE, AMITY, EMOTE, AZTEC, AMMETER'S, GAMETIC, ACETIC, AMIDES, EMOTED, EMOTES, AMIDE'S, AMITY'S
537:& Powervar 14 15: Power var, Power-var, PowerVS, Powervalue, Powers, Power, Power's, However, Powerdev, PowerOff, Powerful, Poweroff, Perv, Weaver
684:& UPM 12 24: UMP, PM, UM, UP, APM, UPI, PPM, WPM, CPM, UPC, UPS, RPM
----- ^^^
````

File: drivers/powervar-hid.c
Comment (aquette) [reply drivers/powervar-hid.c]: thx @andrewmccartney

few more comments in light of this (@jimklimov  @clepple  and others may complete):

* battery.capacitymode: 2
** capacitymode doesn't exist (as per https://github.com/networkupstools/nut/blob/master/docs/nut-names.txt#L350)
** battery.capacity.mode would be more appropriate at first sight (new var, need RFC), but depends on what the values meaning
** for the value, there should be something more friendly to provide (2 == "some explicit string")

* battery.designcapacity
** battery.charge.design would be more appropriate (new var, need RFC), considering the value

* battery.ioeminformation
** non existing variable. battery.mfr or device.mfr.oem or ?

* device.model: UPM UPS (v01.51, Nov 3 2018)
** the part between parenthesis should be split between device.firmware and ups.mfr.date

* ups.powersummary.imanufacturer & ups.powersummary.iproduct & ups.powersummary.iserialnumber
** to be removed, as per my previous comment. these are just the indexes of the strings, already published as device.{mfr, model, serial}

* side comment: I see no load, realpower nor power data, that would give visibility on the power consumption on the UPS output. Nothing avail?

cheers,
Arno

Comment (andrewmccartney) [reply drivers/powervar-hid.c]: @aquette @jimklimov with regards to these changes, if I am understanding you correctly I would need to change the Powervar UPS HID behavior to accomplish this. This would be a UPS firmware change and at this time that is not a possibility. Please advise me if I am not understanding what you are asking here.

**Comment by andrewmccartney** [comment]
@jimklimov I'm new to the process here for merging changes. At this stage the changes are approved - correct? is there anything else I need to do to have them merged?

File: drivers/powervar-hid.c
Comment (aquette) [reply drivers/powervar-hid.c]: @andrewmccartney no need to modify your FW for the above (possibly apart from the realpower, power and load Usage.
The other comments need just the NUT subdriver modification.
device.firmware could also be in its own "UPS.PowerSummary.iVersion", but can also be split from iModel using the callback function:
* modify powervar_format_model() so that it strips the parenthesis and just return the model
* leave a mapping on iProduct, but for device.firmware and set a callback function to extract the firmware info
{ "device.firmware", 0, 0, "UPS.PowerSummary.iProduct", NULL, "%.0f", 0, stringid_to_firmware},
* same for ups.mfr.date...

**Comment by jimklimov** [comment]
Looks good to me, if other NUT maintainers do not voice objections I plan to merge this PR in a week or so.

Thank you very much for the contribution, and for patience with us :) Much appreciated!

File: docs/acknowledgements.txt
Comment (jimklimov) [inline docs/acknowledgements.txt:94]: Note @andrewmccartney : I took the liberty to stub an acknowledgements entry here. Feel free to modify if I got some names or spellings wrong.

**Comment by jimklimov** [comment]
Since other NUT maintainers did not voice objections, I plan to merge this after the build completes which re-syncs the PR code to current master branch state.

**Pull request merged**
